src/Compiler/Microsoft.CodeAnalysis.Razor.Compiler/src/Language/Legacy/RoslynCSharpTokenizer.cs

```diff
@@ -386,404 +210,472 @@ private StateResult Data()
+        var curPosition = Source.Position;
+        var trivia = triviaEnumerator.Current;
+        var triviaString = trivia.ToFullString();
 
-    private static readonly Func<char, bool> IsEndQuotedCharacterLiteral = static (c) => c == '\\' || c == '\'' || SyntaxFacts.IsNewLine(c);
-    private static readonly Func<char, bool> IsEndQuotedStringLiteral = static (c) => c == '\\' || c == '\"' || SyntaxFacts.IsNewLine(c);
-
-    private StateResult QuotedLiteral(char quote, Func<char, bool> isEndQuotedLiteral, SyntaxKind literalType)
-    {
-        TakeUntil(isEndQuotedLiteral);
-        if (CurrentCharacter == '\\')
+        // We handle razor comments with dedicated nodes
+        if (trivia.IsKind(CSharpSyntaxKind.MultiLineCommentTrivia) && triviaString.StartsWith("@*", StringComparison.Ordinal))
         {
-            TakeCurrent(); // Take the '\'
+            Debug.Assert(CurrentCharacter == '@');
+            TakeCurrent();
 
-            // If the next char is the same quote that started this
-            if (CurrentCharacter == quote || CurrentCharacter == '\\')
-            {
-                TakeCurrent(); // Take it so that we don't prematurely end the literal.
-            }
-            return Stay();
-        }
-        else if (EndOfFile || SyntaxFacts.IsNewLine(CurrentCharacter))
-        {
-            CurrentErrors.Add(
-                RazorDiagnosticFactory.CreateParsing_UnterminatedStringLiteral(
-                    new SourceSpan(CurrentStart, contentLength: 1 /* " */)));
+            return Transition(
+                RoslynCSharpTokenizerState.OnRazorCommentStar,
+                EndToken(SyntaxKind.RazorCommentTransition));
         }
-        else
+
+        // Use FullSpan here because doc comment trivias exclude the leading `///` or `/**` and the trailing `*/`
+        var finalPosition = curPosition + trivia.FullSpan.Length;
+
+        for (; curPosition < finalPosition; curPosition++)
         {
-            TakeCurrent(); // No-op if at EOF
+            TakeCurrent();
         }
-        return Transition(CSharpTokenizerState.Data, EndToken(literalType));
-    }
 
-    // CSharp Spec §2.3.2
-    private StateResult BlockComment()
-    {
-        TakeUntil(c => c == '*');
-        if (EndOfFile)
+        if (EndOfFile
+            && trivia.Kind() is CSharpSyntaxKind.MultiLineCommentTrivia or CSharpSyntaxKind.MultiLineDocumentationCommentTrivia
+            && !triviaString.EndsWith("*/", StringComparison.Ordinal))
         {
             CurrentErrors.Add(
                 RazorDiagnosticFactory.CreateParsing_BlockCommentNotTerminated(
                     new SourceSpan(CurrentStart, contentLength: 1 /* end of file */)));
-
-            return Transition(CSharpTokenizerState.Data, EndToken(SyntaxKind.CSharpComment));
         }
-        if (CurrentCharacter == '*')
+
+        // TODO: Handle preprocessor directives
```

Comment:
Should we use PROTOTYPE instead of TODO? Do we have a CI check for any of those words?

Reply:
Yeah, this should be prototype. I don't believe we have any CI for that type of thing, since we don't make feature branches nearly as often as roslyn does.